src/ProductConstructionService/ProductConstructionService.WorkItems/WorkItemScopeManager.cs

```diff
@@ -41,7 +45,11 @@ public async Task<WorkItemScope> BeginWorkItemScopeWhenReadyAsync()
 
     private async Task WorkItemFinishedAsync()
     {
-        await _state.SetStoppedIfStoppingAsync();
+        Interlocked.Decrement(ref _activeWorkItems);
+        if (_activeWorkItems == 0)
+        {
+            await _state.SetStoppedIfStoppingAsync();
```

Comment:
Think we'll need one more change here. Since we only set the state to stopped when all consumers finish their current workitem, we'll have to make this place not enter on `Stopping`,  since that's the state the processor will be in while it's waiting for all Consumers to finish

Reply:
Can you please expand on what you mean here? Not really sure what you mean here

Reply:
With your PR, we only call `SetStoppedWhenStoppingAsync` when we finish the last `WorkItem`. But the method that creates work item scope has this `status == Stopped` as it's condition for waiting. So it will keep consuming new WorkItems, and it's possible that one of the listeners will always get work while a different one is processing it, so we'll never be `Stopped`

Reply:
I think it should just be wait if `status != Working`, it's wrong now too, it will take workitems while it's initializing, I made a bug there too that we just hit

Reply:
So if we receive 3 work items, 3 processors would start. Say, 2 finish early and only the last one is processing.

Currently the other two would just continue pulling items from the queue, which is what we would expect, no? And they can't take the item of the 3rd worker that is being processed as its not in the queue anyways.

Or am I misunderstanding you

Reply:
Yes that is true, but let's say two finish, third one is still working and we do a deployment so the status changes to `Stopping`. In that case we'd wait for someone to call `_state.SetStoppedIfStoppingAsync()` to set the status to `Stopped`. The issue is it's possible for this method to never get called because it's possible `_activeWorkItems` is never `0`. This is because currently, the State check only blocks if the `state = Stopped`. In this case we'd be in `State = Stopping` so we'd keep letting work through, the `_activeWorkItems` would never be 0, so we'd never stop

Reply:
@oleksandr-didyk I pushed the change to the branch because we need it to fix some stuff, hope that's ok with you

Reply:
> @oleksandr-didyk I pushed the change to the branch because we need it to fix some stuff, hope that's ok with you

Of course it is, no worries. Was out unexpectedly anyways